### PR TITLE
preventing zero-length string value for option --limit

### PR DESCRIPTION
**Pull request by viky600833**
##### SUMMARY

When running `ansible-playbook` with an zero length string value for the option `--limit `, the playbook will be applied on whole inventory or whole group specified by `- hosts`
It could result irreparable harm. It should be strictly forbidden.  

The empty argument could occur when running playbook inside a shell script where the shell variable is not valued.
 
  `ansible-playbook test.yaml --limit $testing_group ( $test_group is empty or not setted)`

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

function ansible.cli.CLI.parse(self)

##### ADDITIONAL INFORMATION

* The testing playbook:

```
---
- hosts: all
  gather_facts: no
  tasks:
  - ping:
```

* run playbook as:

```
$ ansible-playbook -i 'centos2,centos3,' test.yaml  --limit ""

PLAY [all] *******************************************************************************************

TASK [ping] ******************************************************************************************
ok: [centos3]
ok: [centos2]

PLAY RECAP *******************************************************************************************
centos2                    : ok=1    changed=0    unreachable=0    failed=0    skipped=0
centos3                    : ok=1    changed=0    unreachable=0    failed=0    skipped=0
```

* in order to prevent it,  non-zero length string check is done in ansible.cli.CLI.parse

* re-run playbook

```
$ ansible-playbook  -i 'centos2,centos3' test.yaml --limit ""
ERROR! Unexpected Exception, this is probably a bug: ansible-playbook: error: --limit option requires non-zero length string
to see the full traceback, use -vvv
```

### CODE REVIEW

File: lib/ansible/cli/__init__.py [hunk 1]
```diff
@@ -575,6 +575,13 @@ def parse(self):
 
         self.options, self.args = self.parser.parse_args(self.args[1:])
 
+        # --limit must receive non-zero length string
+        if hasattr(self.options, 'subset') and  \
+           self.options.subset != C.DEFAULT_SUBSET and \
+           isinstance(self.options.subset, string_types) and \
+           len(self.options.subset) == 0:
```
Comment (sivel) [inline lib/ansible/cli/__init__.py]: Instead of indenting these lines to align with `hasattr` indent them 1 full indentation further than the next line.

```
        if hasattr(self.options, 'subset') and  \
                self.options.subset != C.DEFAULT_SUBSET and \
                isinstance(self.options.subset, string_types) and \
                len(self.options.subset) == 0:
            raise ...
```

File: lib/ansible/cli/__init__.py [hunk 2]
```diff
@@ -575,6 +575,13 @@ def parse(self):
...
+            raise ValueError("ansible-playbook: error: --limit option requires non-zero length string")
```
Comment (sivel) [inline lib/ansible/cli/__init__.py]: You should raise `AnsibleOptionsError` instead of `ValueError`.  Additionally, drop the `ansible-playbook: error: --` and maybe reword to be "Limit requires a non-zero length string"

**Review by sivel — CHANGES_REQUESTED** [review]
This also needs a changelog fragment added in the `changelog/fragments` directory.  See other files for an example.

File: lib/ansible/cli/__init__.py (hunk 2 above)
Comment (viky600833) [reply lib/ansible/cli/__init__.py]: Changes are done in 2 new commit. but CI failed for unknown reason

File: lib/ansible/cli/__init__.py (hunk 1 above)
Comment (viky600833) [reply lib/ansible/cli/__init__.py]: it's done

File: lib/ansible/cli/__init__.py [hunk 3]
```diff
@@ -575,6 +575,13 @@ def parse(self):
 
         self.options, self.args = self.parser.parse_args(self.args[1:])
 
+        # --limit must receive non-zero length string
+        if hasattr(self.options, 'subset') and  \
+                self.options.subset != C.DEFAULT_SUBSET and \
+                isinstance(self.options.subset, string_types) and \
+                len(self.options.subset) == 0:
+            raise AnsibleOptionsError("Limit option requires non-zero length string")
+
```
Comment (bcoca) [inline lib/ansible/cli/__init__.py:582]: I might go for a simpler expression:
```python
limit = getattr(self.options, 'subset', None)
if limit is not None and not limit and limit != C.DEFAULT_SUBSET:
```

Comment (viky600833) [reply lib/ansible/cli/__init__.py:582]: In ansible.cli.CLI.base_parser, OptParser has : add_option('-l','-limit'...default=C.DEFAULT_SUBSET...
I prefer:
limit = getattr(self.options, 'subset', C.DEFAULT_SUBSET)
if limit != C.DEFAULT_SUBSET and not limt: ( this excludes : "", [], {}, None and -1)

**Comment by viky600833** [comment]
It seems something going wrong with osx pipeline

**Comment by samdoran** [comment]
After discussing further, we decided that passing `--limit ''` should run on all hosts.

To discuss this further, please reach out on IRC or the mailing list:

   * IRC: #ansible-devel on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-devel
